### PR TITLE
Improve reconciliation of extension `BackupBucket` and `BackupEntry`

### LINKED ISSUE #7276: `BackupEntry` issues with credential rotation
**How to categorize this issue?**

/area control-plane
/area security
/kind bug

**What happened**:
The backup credentials for a seed cluster were rotated but after the old account was deactivated, some shoots faced issues because the `etcd-backup-restore` side-car still used the expired account.

```
time="2022-12-29T15:10:12Z" level=error msg="Failed initialization: error while initializing: unable to get snapshots from store: Get \"https://www.googleapis.com/storage/v1/b/xyz/o?alt=json&delimiter=&pageToken=&prefix=foo%2Fetcd-main&prettyPrint=false&projection=full&versions=false\": oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error\":\"invalid_grant\",\"error_description\":\"Invalid JWT Signature.\"}" actor=backup-restore-server
```

Situation:
- All shoot clusters were already reconciled successfully after the old account was deactivated.
- The `BackupEntry` secrets in the seed already received that updated account information.
- The `etcd-backup` secret in the shoot control plane namespaces still had the **old account information**.

There are some optimizations in https://github.com/gardener/gardener/blob/387a2cca9683ec2ffdaa3cb0e6c67f46613d343a/pkg/gardenlet/controller/backupentry/backupentry/reconciler.go#L130 which let the controller skip the reconciliation of the `BackupEntry` resource in the seed.
Presumably, after the secret was synced [here](https://github.com/gardener/gardener/blob/387a2cca9683ec2ffdaa3cb0e6c67f46613d343a/pkg/gardenlet/controller/backupentry/backupentry/reconciler.go#L168), Gardenlet was restarted and since then [this condition](https://github.com/gardener/gardener/blob/387a2cca9683ec2ffdaa3cb0e6c67f46613d343a/pkg/gardenlet/controller/backupentry/backupentry/reconciler.go#L163) has led to exiting the reconciliation prematurely. 

**What you expected to happen**:
`etcd-main` not to have any issues after backup credentials rotation.

**Environment**:

- Gardener version: `v1.62.0`

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area control-plane
/kind bug

**What this PR does / why we need it**:
#6926 added watch in the `BackupEntry` controller for extension `BackupEntry`, so we added optimizations to reconcile the extension only when the spec has changed or related secret data has been modified. 
But if the secret gets updated but the controller restarts before the extension `BackupEntry` is reconciled, in the next reconciliation, the extension won't be reconciled because now the secret data is same.
We now add a timestamp to the secret on reconciliation and if the extension `lastUpdateTime` is older than this, we reconcile the extension.

**Which issue(s) this PR fixes**:
Fixes #7276

**Special notes for your reviewer**:
/cc @timuthy 

**Release note**:

```bugfix operator
A bug causing the extension `BackupEntry` not to be correctly reconciled in case of secret rotation and a controller restart is now fixed.
```

### CODE REVIEW

**Comment by shafeeqes** [comment]
> @shafeeqes: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-gardener-integration	[27595be](https://github.com/gardener/gardener/commit/27595be8363d82f523f7a29b5e96f3e0954e60a0)	[link](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7281/pull-gardener-integration/1610856889669652480)	true	`/test pull-gardener-integration`
> [Full PR test history](https://prow.gardener.cloud/pr-history?org=gardener&repo=gardener&pr=7281). [Your PR dashboard](https://prow.gardener.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Ashafeeqes). [Command help for this repository](https://prow.gardener.cloud/command-help?repo=gardener%2Fgardener). Please help us cut down on flakes by linking this test failure to an [open flake report](https://github.com/gardener/gardener/issues?q=is:issue+is:open+label:kind/flake) or filing a [new flake report](https://github.com/gardener/gardener/issues/new?template=flaking-test.md&title=%5BFlaky+Test%5D+FLAKING+TEST%2FSUITE) if you can't find an existing one. Also see our [testing guideline](https://github.com/gardener/gardener/blob/master/docs/development/testing.md) for how to avoid and hunt flakes.

https://github.com/gardener/gardener/issues/6933
/test pull-gardener-integration

**Comment by shafeeqes** [comment]
Will manually open a PR for v1.60, since we need to cherry-pick only the backupbucket commit.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go [hunk 1]
```diff
@@ -183,14 +189,20 @@ func (r *Reconciler) reconcileBackupEntry(
 		BackupBucketProviderStatus: backupBucket.Status.ProviderStatus,
 	}
 
+	secretLastUpdateTime, err := time.Parse(time.RFC3339, extensionSecret.Annotations[v1beta1constants.GardenerTimestamp])
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupentry/backupentry/reconciler.go:198]: If we keep `mustReconcileExtensionSecret`:
What happens for existing clusters where the timestamp is not set yet? I think we have to consider this case as well.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler_test.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		Expect(seedClient.Create(ctx, extensionSecret)).To(Succeed())
+		Expect(seedClient.Create(ctx, extensionBackupEntry)).To(Succeed())
+
+		result, err := reconciler.Reconcile(ctx, request)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(result).To(Equal(reconcile.Result{}))
+
+		Expect(seedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupEntry), extensionBackupEntry)).To(Succeed())
+		Expect(extensionBackupEntry.Annotations).To(HaveKeyWithValue(v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationReconcile))
+	})
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupentry/backupentry/reconciler_test.go:270]: See comment above, please also add a test that considers secrets w/o a timestamp being set.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go [hunk 3]
```diff
@@ -128,6 +128,7 @@ func (r *Reconciler) reconcileBackupEntry(
 
 	var (
 		mustReconcileExtensionBackupEntry = false
+		mustReconcileExtensionSecret      = false
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: Can you explain why we need this "optimization"?

File: pkg/gardenlet/controller/backupbucket/reconciler.go [hunk 4]
```diff
@@ -127,6 +127,7 @@ func (r *Reconciler) reconcileBackupBucket(
 
 	var (
 		mustReconcileExtensionBackupBucket = false
+		mustReconcileExtensionSecret       = false
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupbucket/reconciler.go]: Same concerns as for `BackupEntry`.

File: pkg/gardenlet/controller/backupbucket/reconciler.go [hunk 5]
```diff
@@ -234,7 +234,7 @@ func (r *Reconciler) reconcileBackupBucket(
 		return reconcile.Result{}, nil
 	}
 
-	if reconciliationSuccessful {
+	if reconciliationSuccessful && extensionBackupBucket.Status.LastOperation.State == gardencorev1beta1.LastOperationStateSucceeded {
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupbucket/reconciler.go]: Why do we need this second condition? I think we can drop `reconciliationSuccessful` if you prefer the latter condition.

File: pkg/gardenlet/controller/backupbucket/reconciler.go [hunk 6]
```diff
@@ -182,6 +183,7 @@ func (r *Reconciler) reconcileBackupBucket(
 	} else if extensionBackupBucket.Status.LastOperation == nil {
 		// if the extension did not record a lastOperation yet, don't update the status
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupbucket/reconciler.go]: I think with the changed error handling, the status is indeed updated. Can you please update this comment and also did you check if this has some unwanted side effects? There might be an error reported in the `BackupBucket` object in the garden cluster very early now.

**Review by timuthy — COMMENTED** [review]
Thanks a lot for the improvements 🙂

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go [hunk 7]
```diff
@@ -605,10 +617,9 @@ func (r *Reconciler) getGardenSecret(ctx context.Context, backupBucket *gardenco
 	return gardenSecret, nil
 }
 
-func (r *Reconciler) reconcileBackupEntryExtensionSecret(ctx context.Context, backupBucket *gardencorev1beta1.BackupBucket, backupEntry *gardencorev1beta1.BackupEntry, gardenSecret *corev1.Secret) error {
-	// create secret for extension BackupEntry in seed
-	extensionSecret := r.emptyExtensionSecret(backupEntry)
```
Comment (ary1992) [inline pkg/gardenlet/controller/backupentry/backupentry/reconciler.go:610]: Why this statement is removed?

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go (hunk 3 above)
Comment (shafeeqes) [reply pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: Otherwise, the secret will be reconciled everytime, timestamp will be set, and it will be greater than the extension `lastUpdateTime` obviously.
Maybe if we use the generation of secret like you had suggested, this won't happen because the generation won't be updated even if we reconcile the secret and the data hasn't changed. But I felt the timestamp approach is better.

File: pkg/gardenlet/controller/backupbucket/reconciler.go (hunk 6 above)
Comment (shafeeqes) [reply pkg/gardenlet/controller/backupbucket/reconciler.go]: So, ideally, this shouldn't happen because of the predicate here,
https://github.com/gardener/gardener/blob/c11c86ae07d8ea784f5c41362cd41800f06bb3ed/pkg/controllerutils/predicate/predicate.go#L226-L228
But during testing, at some point, there was a nil pointer exception because the extension backupentry status didn't have lastoperation, so I had put this as a safety net.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go (hunk 7 above)
Comment (shafeeqes) [reply pkg/gardenlet/controller/backupentry/backupentry/reconciler.go:610]: Didn't feel it provides much value.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go [hunk 8]
```diff
@@ -222,12 +220,21 @@ func (r *Reconciler) reconcileBackupEntry(
...
+		} else if lastOperationState == gardencorev1beta1.LastOperationStateSucceeded {
+			if updateErr := r.updateBackupEntryStatusSucceeded(ctx, backupEntry, operationType); updateErr != nil {
+				return reconcile.Result{}, fmt.Errorf("could not update status after reconciliation success: %w", updateErr)
+			}
+
+			if kutil.HasMetaDataAnnotation(&backupEntry.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationRestore) {
+				if updateErr := removeGardenerOperationAnnotation(ctx, r.GardenClient, backupEntry); updateErr != nil {
+					return reconcile.Result{}, fmt.Errorf("could not remove %q annotation: %w", v1beta1constants.GardenerOperation, updateErr)
+				}
+			}
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: Isn't the `BackupBucket` overall updated to early with this change? Let's suppose an issues occurs later in `reconcileBackupEntryExtension`. Then the successful operation is already published even though the reconciliation failed.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go (hunk 3 above)
Comment (timuthy) [reply pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: Okay, got it. Maybe you can add this as a comment that this really must happen on purpose.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go (hunk 8 above)
Comment (shafeeqes) [reply pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: In all cases where we set `mustReconcileExtensionBackupEntry = true`, we are not reaching this code.
And if the error occurs in the extension backupbucket in the next reconciliation, the reconciler will be invoked when it's status is updated, where we will capture this error. That's the reason why we exit reconciliation early here, https://github.com/gardener/gardener/blob/4a545ac112d135de9f3bce8a4cc66db9037de92b/pkg/gardenlet/controller/backupentry/backupentry/reconciler.go#L259-L260 too.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go (hunk 3 above)
Comment (shafeeqes) [reply pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: Done in https://github.com/gardener/gardener/pull/7281/commits/71f48b8b32288ec421b603be7ad55200ef094776

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler.go (hunk 8 above)
Comment (timuthy) [reply pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: Got it, thanks 🙂 

TBH, I don't like this structure too much because there is a lot of nesting esp. here where the "final step" of the reconciliation (updating `BackupBucket` to succeeded) is somewhere in between. Even though it's technically possible, readability suffers.
I think it was more understandable when this "final step" was an extra code block at the end.

Comment (shafeeqes) [reply pkg/gardenlet/controller/backupentry/backupentry/reconciler.go]: Agreed, PTAL: https://github.com/gardener/gardener/pull/7281/commits/f310fa1577975f2a246f98923f9384570e36c664

File: test/integration/gardenlet/backupentry/backupentry/backupentry_suite_test.go [hunk 9]
```diff
@@ -140,7 +140,7 @@ var _ = BeforeSuite(func() {
 		},
 	}
 
-	Expect(testClient.Create(ctx, gardenNamespace)).To(Or(Succeed(), BeAlreadyExistsError()))
```
Comment (timuthy) [inline test/integration/gardenlet/backupentry/backupentry/backupentry_suite_test.go:143]: Why is this changed? Shouldn't we tolerate this error in case the integration test runs against an existing cluster?

I have to admit, even this is odd because we delete the namespace again after test execution. However, we follow this logic in several integration tests and should stick to either of the approaches.

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler_test.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+		result, err := reconciler.Reconcile(ctx, request)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(result).To(Equal(reconcile.Result{}))
+
+		Expect(seedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupEntry), extensionBackupEntry)).To(Succeed())
+		Expect(extensionBackupEntry.Annotations).NotTo(HaveKey(v1beta1constants.GardenerOperation))
+	})
+
+	It("should reconcile the extension secret and extension backupentry if the secret currently doesn't have a timestamp", func() {
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupentry/backupentry/reconciler_test.go]:
```suggestion
	It("should reconcile the extension secret and extension BackupEntry if the secret currently doesn't have a timestamp", func() {
```

File: pkg/gardenlet/controller/backupbucket/reconciler_test.go [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+		result, err := reconciler.Reconcile(ctx, request)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(result).To(Equal(reconcile.Result{}))
+
+		Expect(seedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupBucket), extensionBackupBucket)).To(Succeed())
+		Expect(extensionBackupBucket.Annotations).To(HaveKeyWithValue(v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationReconcile))
+	})
+
+	It("should reconcile the extension BackupBucket if the secret update timestamp is after the extension lastupdate time", func() {
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupbucket/reconciler_test.go]:
```suggestion
	It("should reconcile the extension BackupBucket if the secret update timestamp is after the extension last update time", func() {
```

File: pkg/gardenlet/controller/backupbucket/reconciler_test.go [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+		result, err := reconciler.Reconcile(ctx, request)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(result).To(Equal(reconcile.Result{}))
+
+		Expect(seedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupBucket), extensionBackupBucket)).To(Succeed())
+		Expect(extensionBackupBucket.Annotations).NotTo(HaveKey(v1beta1constants.GardenerOperation))
+	})
+
+	It("should reconcile the extension secret and extension backupentry if the secret currently doesn't have a timestamp", func() {
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupbucket/reconciler_test.go]:
```suggestion
	It("should reconcile the extension secret and extension BackupBucket if the secret currently doesn't have a timestamp", func() {
```

File: pkg/gardenlet/controller/backupentry/backupentry/reconciler_test.go [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+
+		result, err := reconciler.Reconcile(ctx, request)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(result).To(Equal(reconcile.Result{}))
+
+		Expect(seedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupEntry), extensionBackupEntry)).To(Succeed())
+		Expect(extensionBackupEntry.Annotations).To(HaveKeyWithValue(v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationReconcile))
+	})
+
+	It("should reconcile the extension BackupEntry if the secret update timestamp is after the extension lastupdate time", func() {
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupentry/backupentry/reconciler_test.go]:
```suggestion
	It("should reconcile the extension BackupEntry if the secret update timestamp is after the extension last update time", func() {
```

**Review by timuthy — COMMENTED** [review]
Some last nits, the rest looks good to me.
Thanks for taking care!

File: test/integration/gardenlet/backupentry/backupentry/backupentry_suite_test.go (hunk 9 above)
Comment (shafeeqes) [reply test/integration/gardenlet/backupentry/backupentry/backupentry_suite_test.go:143]: From https://github.com/gardener/gardener/blob/master/docs/development/testing.md#writing-integration-tests,
> don't tolerate already existing resources (~dirty test environment), code smell: ignoring already exist errors

>  However, we follow this logic in several integration tests and should stick to either of the approaches.

I agree, we should streamline this. I can do it as a separate PR.

Comment (shafeeqes) [reply test/integration/gardenlet/backupentry/backupentry/backupentry_suite_test.go:143]: Opened https://github.com/gardener/gardener/pull/7329

**Review by timuthy — COMMENTED** [review]
/lgtm
